### PR TITLE
Bug fix for tiny edge-case in MarkDuplicates

### PR DESCRIPTION
**Pull request by yfarjoun**
- Fixed an edge-case that had non-consistent behavior. Added a test that covers it.
- Removed a redundant call to "buildReadEnds" by cloning the fragmentEnd instead.
- Clarified code by changing some variable names.

@nh13 @tfenne

### CODE REVIEW

File: src/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 1]
```diff
@@ -461,24 +463,33 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
...
                                     pairedEnds.orientation == ReadEnds.R);
                         }
+                        // if the two read ends are in the same position, pointing in opposite directions,
+                        // the orientation is undefined and the procedure above
+                        // will depend on the order of the reads in the file.
+                        // To avoid this, we set it explicitly (to FR):
+                        if (pairedEnds.read2ReferenceIndex == pairedEnds.read1ReferenceIndex &&
+                                pairedEnds.read2Coordinate == pairedEnds.read1Coordinate &&
+                                pairedEnds.orientation == ReadEnds.RF) {
+                            pairedEnds.orientation = ReadEnds.FR;
```
Comment (nh13) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: I think this is confusing to be separate from the `if/else` starting on line 467.  Can you make it part of the first `if` on line 467, or make it its own `if (sequence == pairedEnds.read1ReferenceIndex && coordinate == pairedEnds.read1Coordinate)` and then you can have `if/else if/else`?  I would prefer the former.

File: src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicates.java [hunk 2]
```diff
@@ -71,4 +71,9 @@ public ReadEndsForMarkDuplicates(final ReadEndsForMarkDuplicates read) {
         this.read1IndexInFile = read.read1IndexInFile;
         this.read2IndexInFile = read.read2IndexInFile;
     }
+
+    @Override
+    public ReadEndsForMarkDuplicates clone() {
```
Comment (nh13) [inline src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicates.java]: I'd actually prefer a copy constructor or method over a clone method.

File: src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicatesWithBarcodes.java [hunk 3]
```diff
@@ -24,7 +24,7 @@
 
 package picard.sam.markduplicates.util;
 
-public class ReadEndsForMarkDuplicatesWithBarcodes extends ReadEndsForMarkDuplicates {
+public class ReadEndsForMarkDuplicatesWithBarcodes extends ReadEndsForMarkDuplicates  {
```
Comment (nh13) [inline src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicatesWithBarcodes.java]: I see a spaces.

File: src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java [hunk 4]
```diff
@@ -42,4 +54,41 @@ protected boolean markUnmappedRecordsLikeTheirMates() {
...
+        }
+
+        @Override
+        protected CommandLineProgram getProgram() {
+            return new MarkDuplicates();
+        }
+    }
+
+    @DataProvider
+    public Object[][] SameUnclipped5primeEnd() {
```
Comment (nh13) [inline src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java]: methods should start with a lower case, how about `testStrangeOrderDependentBugVersionDataProvider`?

File: src/tests/java/picard/sam/testers/SamFileTester.java [hunk 5]
```diff
@@ -340,4 +342,8 @@ private File createInputFile() {
     public SamReader getInput() {
         return samRecordSetBuilder.getSamReader();
     }
+
+    public void setRecordsNeedSorting(final boolean recordsNeedSorting) {
```
Comment (nh13) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: could this return the instance of `SamFileTester` to allow chaining?  Then the name could be `recordsNeedSorting(...)`.

File: src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java [hunk 6]
```diff
@@ -42,4 +54,41 @@ protected boolean markUnmappedRecordsLikeTheirMates() {
...
+
+            new Object[]{new File(TEST_DIR, "SameUnclipped5primeEndv1.sam")},
+            new Object[]{new File(TEST_DIR, "SameUnclipped5primeEndv2.sam")}};
+    }
+
+    @Test(dataProvider = "SameUnclipped5primeEnd")
+    public void testStrangeOrderDependentBugVersion(final File input) {
+
+        final AbstractMarkDuplicatesCommandLineProgramTester tester = new ByQualityQuerySortedTester();
+        tester.setRecordsNeedSorting(false);
```
Comment (nh13) [inline src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java]: could chain this if `setRecordsNeedSorting` returned the tester.

File: testdata/picard/sam/MarkDuplicates/SameUnclipped5primeEndv1.sam [hunk 7]
```diff
@@ -0,0 +1,28 @@
+@HD	VN:1.5	SO:queryname
```
Comment (nh13) [inline testdata/picard/sam/MarkDuplicates/SameUnclipped5primeEndv1.sam]: most other files are either lower case with underscore separators, or camel-case with a lower case first character.

**Comment by nh13** [comment]
@yfarjoun minor comments.  I had to convince myself this was the right thing to do, and not dependent on which read was the first in a pair, which matters for optical duplicate detection.

File: src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicatesWithBarcodes.java (hunk 3 above)
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicatesWithBarcodes.java]: I left you an easter egg :-)

File: src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicates.java (hunk 2 above)
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicates.java]: That was my initial preference as well but I couldn't make it work. the problem is that I do not know the actual type of the object since it could be either a ReadEndsForMarkDuplicates or a ReadEndsForMarkDuplicatesWithBarcodes. if there's a way to do it without a clone, let me know.

File: src/java/picard/sam/markduplicates/MarkDuplicates.java (hunk 1 above)
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/MarkDuplicates.java]: The former will mean moving the if in its entirety into the first if, there is (almost) no saving on the logic..it that OK?

File: src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java (hunk 4 above)
Comment (yfarjoun) [inline src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java]: I should ajust the name...it was named before I understood the problem 😄

File: src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java (hunk 6 above)
Comment (yfarjoun) [inline src/tests/java/picard/sam/markduplicates/QuerySortedMarkDuplicatesTest.java]: How does that work in class inheritance? What would be the signature of setRecordsNeedSorting?

File: src/tests/java/picard/sam/testers/SamFileTester.java (hunk 5 above)
Comment (yfarjoun) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: but I need an AbstractMarkDuplicatesCommandLineProgramTester instance...

File: testdata/picard/sam/MarkDuplicates/SameUnclipped5primeEndv1.sam (hunk 7 above)
Comment (yfarjoun) [inline testdata/picard/sam/MarkDuplicates/SameUnclipped5primeEndv1.sam]: I got messed up with the capitalization...I'll fix when I squash.

File: src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicates.java (hunk 2 above)
Comment (nh13) [inline src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicates.java]: Your call, but consider adding `implements Cloneable` to each class that implements `clone`.

File: src/tests/java/picard/sam/testers/SamFileTester.java (hunk 5 above)
Comment (nh13) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: Not the most beautiful, so perhaps it is not needed
```
    @SuppressWarnings("unchecked")
    public <T extends SamFileTester> T setRecordsNeedSorting(final boolean recordsNeedSorting) {
        this.recordsNeedSorting = recordsNeedSorting;
        return (T) this;
    }
```

**Comment by nh13** [comment]
@yfarjoun rebase and pass the tests.  Your call on the additional comments I made.

File: src/tests/java/picard/sam/testers/SamFileTester.java (hunk 5 above)
Comment (yfarjoun) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: I am having a hard time deciding what's a better style to have....sage advice?

File: src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicates.java (hunk 2 above)
Comment (yfarjoun) [inline src/java/picard/sam/markduplicates/util/ReadEndsForMarkDuplicates.java]: does a subclass also need `implements Clonable`?

File: src/tests/java/picard/sam/testers/SamFileTester.java (hunk 5 above)
Comment (nh13) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: Bringing in @tfenne 
top-gun-switching-to-guns

Comment (tfenne) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: So I think the "right" option here is also fairly ugly, so I'll leave it up to you guys whether or not you want to implement it.  It would require declaring `SamFileTester` like this:

```java
class SamFileTester<T extends SamFileTester<T>> {
    public T setRecordsNeedSorting(final boolean recordsNeedSorting) {
        this.recordsNeedSorting = recordsNeedSorting;
        return (T) this;
    }
}
```

And then having subclasses do things like:

```java
class MdSamFileTester extends SamFileTester<MdSamFileTester> { ... }
```

This also means that instantiating the base class becomes ugly: `new SamFileTester<SamFileTester>()`.  You could of course overcome this by providing a static factory method, or having an abstract base class with the generic crap in it and an empty extension class that's instantiated.

**Comment by yfarjoun** [comment]
I think that it only come into play in querysorted bams because in coordinate sorted bams if the 5' ends are the same and the strands are opposite, the alignment start can be the same only if one of the reads is almost entirely clipped (except one base). I'll add that as another crazy test-case....

File: src/tests/java/picard/sam/testers/SamFileTester.java (hunk 5 above)
Comment (yfarjoun) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: wouldn't it be
 `MdSamFileTester = new MdSamFileTester <>()` 
in java8?
not too bad...

Comment (tfenne) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: If that works without generating warnings, then yes.

Comment (yfarjoun) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: bah. all for chaining in a test framework...I'm going to punt on this...

Comment (nh13) [inline src/tests/java/picard/sam/testers/SamFileTester.java]: 👍 TIL, but punt it.
